presto-main/src/main/java/io/prestosql/execution/StageState.java

```diff
@@ -88,4 +88,29 @@ public boolean isFailure()
     {
         return failureState;
     }
+
+    public boolean noMoreTasks()
+    {
+        boolean noMoreTasks = false;
+        switch (this) {
+            case PLANNED:
+            case SCHEDULING:
+                // workers are still being added to the query
+                break;
+            case SCHEDULING_SPLITS:
+            case SCHEDULED:
+            case RUNNING:
+            case FINISHED:
+            case CANCELED:
+                // no more workers will be added to the query
```

Comment:
The comments don't seem to match the code... this doesn't change anything

Reply:
This code was extracted as-is from `io.prestosql.execution.scheduler.SqlQueryScheduler.StageLinkage#processScheduleResults`
https://github.com/prestosql/presto/pull/1072#discussion_r370150845
Are we guaranteed that there will be no more tasks added (e.g when new workers come with co-located joins) when we reach SCHEDULING_SPLITS state ? 
If not, we still need some way to figure out in `DynamicFilterService` when it's safe to assume that no new tasks will be added which can generate more values from build-side of join and we can use the dynamic filters reported from completion of existing build-side operators for filtering on the probe side.

Reply:
@dain 

Looking at `SourcePartitionedScheduler#finalizeTaskCreationIfNecessary` it will `splitPlacementPolicy.lockDownNodes();` and set state to `SCHEDULING_SPLIT`. It seems that at this point no more nodes will participate in query.

However, can stage state can change to `RUNNING` via `SqlStageExecution#schedulingComplete()` and skip `splitPlacementPolicy.lockDownNodes();` call?

I guess `SqlQueryScheduler.StageLinkage#processScheduleResults` would also break then